src/typechecker.erl

```diff
@@ -4165,87 +4217,95 @@ add_type_pat(Pat, ?type(union, UnionTys)=UnionTy, Env, VEnv) ->
+            NewEnv = union_var_binds([ add_any_types_pat(Pat, Env) || Pat <- Pats ], Env),
+            {type(none),
+             Ty,
+             NewEnv,
+             constraints:empty()};
         {elem_ty, Tys, Cs} ->
-            {PatTys, UBounds, VEnv1, Cs1} = do_add_types_pats(Pats, Tys, Env, VEnv),
-            {type(tuple, PatTys)
-            ,type(tuple, UBounds)
-            ,VEnv1
-            ,constraints:combine(Cs, Cs1)};
+            {PatTys, UBounds, Env1, Cs1} = do_add_types_pats(Pats, Tys, Env),
+            {type(tuple, PatTys),
+             type(tuple, UBounds),
+             Env1,
+             constraints:combine(Cs, Cs1)};
         {type_error, _Type} ->
             throw({type_error, pattern, P, Tuple, Ty})
     end;
-add_type_pat(Atom = {atom, P, Val}, Ty, Env, VEnv) ->
+add_type_pat(Atom = {atom, P, Val}, Ty, Env) ->
     LitTy = {atom, erl_anno:new(0), Val},
     case subtype(LitTy, Ty, Env) of
         {true, Cs} ->
-            {LitTy, LitTy, VEnv, Cs};
+            {LitTy, LitTy, Env, Cs};
         false ->
             throw({type_error, pattern, P, Atom, Ty})
     end;
-add_type_pat(Nil = {nil, P}, Ty, Env, VEnv) ->
+add_type_pat(Nil = {nil, P}, Ty, Env) ->
     NilTy = type(nil),
     case subtype(NilTy, Ty, Env) of
         {true, Cs} ->
-            {NilTy, NilTy, VEnv, Cs};
+            {NilTy, NilTy, Env, Cs};
         false ->
             throw({type_error, pattern, P, Nil, Ty})
     end;
-add_type_pat(CONS = {cons, P, PH, PT}, ListTy, Env, VEnv) ->
+add_type_pat(CONS = {cons, P, PH, PT}, ListTy, Env) ->
     case expect_list_type(normalize(ListTy, Env), dont_allow_nil_type, Env) of
         any ->
-            VEnv2 = add_any_types_pat(PH, VEnv),
+            Env2 = add_any_types_pat(PH, Env),
             TailTy = normalize(type(union, [ListTy, type(nil)]), Env),
-            {_TailPatTy, _TauUBound, VEnv3, Cs} = add_type_pat(PT, TailTy, Env, VEnv2),
+            {_TailPatTy, _TauUBound, VEnv3, Cs} = add_type_pat(PT, TailTy, Env2),
             NonEmptyTy = rewrite_list_to_nonempty_list(ListTy),
-            {NonEmptyTy, NonEmptyTy, VEnv3, Cs};
+            {type(none), NonEmptyTy, VEnv3, Cs};
         {elem_ty, ElemTy, Cs1} ->
-            {_PatTy1, _UBound1, VEnv2, Cs2} =
-                add_type_pat(PH, normalize(ElemTy, Env), Env, VEnv),
+            {PatTy1, _UBound1, Env2, Cs2} =
+                add_type_pat(PH, normalize(ElemTy, Env), Env),
             TailTy = normalize(type(union, [ListTy, type(nil)]), Env),
-            {_PatTy2, _Ubound2, VEnv3, Cs3} = add_type_pat(PT, TailTy, Env, VEnv2),
+            {_PatTy2, _Ubound2, Env3, Cs3} = add_type_pat(PT, TailTy, Env2),
+            {PatTy, Env4} = case is_list_pat_exhaustive(CONS) of
+                                true ->
+                                    {type(nonempty_list, [PatTy1]), Env3};
+                                false ->
+                                    {type(none), Env3#env{exhaust = false}}
```

Comment:
Where do you set exhaust back to what it was before? Does this disable exhaustiveness checking for the rest of the module?

In a nested pattern (or nested expression) think we want to disable exhaustiveness checking only for the particular list where we don't have an exhaustive pattern in the clauses. Below, we still want it for the tuple and for ANotherList, don't we?

```Erlang
case List of
    {[X, Y | Tail], foo} ->
        case ANotherList of [] -> empty; [_|_] -> nonempty end;
    {[X], foo} ->
        ...
    {[], foo} ->
        ...
end
```

Reply:
> Where do you set exhaust back to what it was before? Does this disable exhaustiveness checking for the rest of the module?

That's a bit tricky. I don't yet, but the next form will be checked with a different copy of `Env` where `exhaust` is `true` by default:
https://github.com/josefs/Gradualizer/blob/3cda2e5aa1b1313b3e02fce67986fcda3ab9a02c/src/typechecker.erl#L4892
In other words, the current toggle disables exhaustiveness checking "for the rest of the current top-level form".

Based on your example, if we toggle `exhaust` to `true` when leaving `check_clauses`:

```erlang
case List of
    {[X, Y | Tail], foo} -> %% we set exhaust=false here
        case ANotherList of %% we won't check exhaustiveness on the nested case expr
            [] -> empty;
            [_|_] -> nonempty
        end; %% we exit from check_clauses -> set exhaust=true
    {[X], foo} -> %% we have exhaust=true, even though we wanted exhaust=false and did set it on the previous clause of this case expr
        ...
    {[], foo} ->
        ...
end
```

We could probably keep a stack of `exhaust` values in `Env` and push the previous value when entering `check_clauses` and pop when exiting 🤔  WDYT?

Reply:
Stack! :+1: :100: I was thinking about that too. We don't need to change exhaust=true (the user's option) but instead we can use a different field (`clause_stack = [#clause{exhaust=true}|Clauses]`). WDYT?

Reply:
OK, this is done now and seems to work well :+1: